idea/src/org/jetbrains/kotlin/idea/intentions/AddLabeledReturnInLambdaIntention.kt

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright 2000-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.intentions
+
+import com.intellij.codeInsight.intention.LowPriorityAction
+import com.intellij.openapi.editor.Editor
+import org.jetbrains.kotlin.idea.caches.resolve.analyze
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.getCallNameExpression
+import org.jetbrains.kotlin.psi.psiUtil.getNonStrictParentOfType
+import org.jetbrains.kotlin.psi.psiUtil.getStrictParentOfType
+import org.jetbrains.kotlin.resolve.calls.callUtil.getResolvedCall
+import org.jetbrains.kotlin.types.typeUtil.isUnit
+
+class AddLabeledReturnInLambdaIntention :
+        SelfTargetingIntention<KtExpression>(KtExpression::class.java, "Add labeled return to last expression in a lambda"),
+        LowPriorityAction {
+    override fun isApplicableTo(element: KtExpression, caretOffset: Int): Boolean {
+        if (element.getNonStrictParentOfType<KtReturnExpression>() != null) return false
+        val block = element.getStrictParentOfType<KtBlockExpression>() ?: return false
+        if (block.statements.lastOrNull() != element) return false
+        val callExpression = element.getStrictParentOfType<KtCallExpression>() ?: return false
+        val index = callExpression.valueArguments.indexOfFirst {
+            val lambda = it.getArgumentExpression() as? KtLambdaExpression
+            lambda?.bodyExpression == block
+        }
+        if (index < 0) return false
+        val resolvedCall = callExpression.getResolvedCall(callExpression.analyze()) ?: return false
+        if (resolvedCall.candidateDescriptor.valueParameters[index].type.arguments.last().type.isUnit()) return false
```

Comment:
Why this case with `Unit` result is important? Why can't we write something like
```
1.let {
    return@let Unit
}
```

Reply:
Anyway, test should be created for this case

Reply:
I don't want to add `return@forEach` in the following case.

```
list(1,2).forEach {
    <caret>1
}
```

```
list(1,2).forEach {
    return@forEach 1 // this would be compile error
}
```

Reply:
@mglukhikh what should I do for these case? Do you want to add `return@forEach` at the end of lambda or ignore the case?

Reply:
@shiraji You can try `KtExpression.isUsedAsExpression` to control whether this `1` is really lambda result or not. In the second case, `isUsedAsExpression` should return false.